_pytest/cacheprovider.py

```diff
@@ -8,13 +8,17 @@
 import py
 import pytest
 import json
-from os.path import sep as _sep, altsep as _altsep
+from os.path import sep as _sep, altsep as _altsep, \
+                    isabs as _isabs, expanduser as _expanduser
 
 
 class Cache(object):
     def __init__(self, config):
         self.config = config
-        self._cachedir = config.rootdir.join(".cache")
+        cache_dirname = config.getini("cache_dirname")
+        if _isabs(_expanduser(cache_dirname)):
+            raise ValueError("cache dirname must be relative path, not absolute")
+        self._cachedir = config.rootdir.join(cache_dirname)
```

Comment:
I suggest to create a staticmethod `Cache.cache_dir_from_config(config)` which contains this logic; this would enhance the readability and also allow you to unittest the logic inside it more easily.